File: engine/roster.py
"""The subscriber's roster, resolved to ids we can actually score.

WHY THIS EXISTS. Sleeper's terms forbid us reading the league (PLAN §0), so the
league half of the product now comes from the only party entitled to hand it
over: the subscriber, typing it. That makes NAME RESOLUTION the correctness
heart of the product. A wrong match does not fail loudly — it produces a
complete, confident report about a player the subscriber does not own, which is
the worst failure this product can have (principle 3).

MEASURED, NOT ASSUMED. Against nflverse's players release (Aug 18 2026):
- 6,079 of 25,040 rows carry a ``gsis_id`` that is NOT GSIS-format (``ABB498348``
  and similar — players with no GSIS id yet). They duplicate real humans: Layne
  Pryor appears as both ``00-0040792`` and ``PRY456541``. **RULE R1: only
  ``00-#######`` ids are eligible.** Skipping this filter silently doubles
  people.
- Restricted to fantasy positions with a GSIS id and ``last_season >= 2024``,
  the pool is 1,327 players with **ZERO normalised-name collisions**. Across all
  time it is 156 collisions (two Adrian Petersons, an Alex Smith at QB and
  another at TE), so **RULE R2: recency is what makes exact matching safe.** The
  window is a parameter, not a constant, because it is the load-bearing
  assumption.
- Name shapes to survive: 64 periods (A.J. Brown), 62 suffixes (Kenneth Walker
  III), 32 apostrophes (Ja'Marr Chase), 23 hyphens (Amon-Ra St. Brown), 1
  non-ASCII.

**RULE R3 — AMBIGUITY IS RETURNED, NEVER RESOLVED.** When a name matches more
than one eligible player, or none, this module says so and names the candidates.
It does not pick. Guessing here is indistinguishable from working right up until
the report goes out, and the person who knows the answer is the one typing.
"""

from __future__ import annotations

import csv
import re
import unicodedata
from dataclasses import dataclass
from pathlib import Path

# RULE R1. Anything else in that column is a player without a GSIS id, and the
# same human usually appears again WITH one.
GSIS_RE = re.compile(r"^00-\d{7}$")

# Positions a fantasy roster can hold. FB is included because leagues that use
# it slot it at RB, and excluding it would make a real roster unresolvable.
FANTASY_POSITIONS = frozenset({"QB", "RB", "WR", "TE", "K", "FB"})

# Suffixes are dropped: managers type "Kenneth Walker" for "Kenneth Walker III"
# far more often than the reverse, and no eligible pair differs ONLY by suffix.
_SUFFIX_RE = re.compile(r"\b(jr|sr|ii|iii|iv|v)\b\.?")

DEFENSE = "DEF"


def normalize(name: str) -> str:
    """Fold a typed name to its comparison form: letters only, no spaces.

    Order matters: accents are stripped before the alphabetic filter (so "José"
    becomes "jose" rather than "jos"), and suffixes go before punctuation (so
    "Jr." is caught while it still has its period).

    **Spaces and punctuation are DELETED, not collapsed.** Keeping them meant
    "JAMARR CHASE" missed "Ja'Marr Chase" and "AJ Brown" missed "A.J. Brown" —
    people do not reproduce apostrophes and periods, and a miss here is a
    subscriber who cannot finish signup. This is safe on measurement, not on
    hope: over the eligible pool (1,327 players) a letters-only key produces
    **zero collisions**, the same as the spaced form. ``test_roster.py`` asserts
    that against the real directory so the day it stops being true is loud.
    """
    folded = unicodedata.normalize("NFKD", name or "")
    folded = "".join(c for c in folded if not unicodedata.combining(c)).lower()
    folded = _SUFFIX_RE.sub(" ", folded)
    return re.sub(r"[^a-z]", "", folded)


@dataclass(frozen=True)
class Player:
    """One rosterable entity: a real player, or a team defense."""

    player_id: str          # GSIS id, or "DEF-BAL" for a team defense
    name: str
    position: str
    team: str | None

    @property
    def is_defense(self) -> bool:
        return self.position == DEFENSE


@dataclass(frozen=True)
class Match:
    """What resolution concluded about one typed line.

    ``player`` is set only when exactly one eligible player matched. Otherwise
    ``candidates`` carries what we found so the human can choose — RULE R3.
    """

    typed: str
    player: Player | None
    candidates: tuple[Player, ...] = ()

    @property
    def resolved(self) -> bool:
        return self.player is not None

    @property
    def reason(self) -> str | None:
        """Buyer-facing, in the report's own register — no ids, no jargon."""
        if self.resolved:
            return None
        if not self.candidates:
            return "we don't have a player by that name"
        return f"more than one player goes by that name ({len(self.candidates)})"


class PlayerDirectory:
    """Every player and defense a subscriber may name, indexed for lookup."""

    def __init__(self, players: list[Player]) -> None:
        self.players = players
        self._teams = {p.team for p in players if p.team}
        self._by_name: dict[str, list[Player]] = {}
        for player in players:
            self._by_name.setdefault(normalize(player.name), []).append(player)
        # Defenses are matched separately: a manager writes them a dozen ways
        # ("Ravens", "BAL DEF", "Baltimore Ravens D/ST"), and none of those is
        # the display name of a person.
        self._defense_alias: dict[str, Player] = {}
        for player in players:
            if player.is_defense and player.team:
                for alias in _defense_aliases(player):
                    self._defense_alias[alias] = player

    def __len__(self) -> int:
        return len(self.players)

    def resolve(self, typed: str) -> Match:
        """One typed line -> a Match. Never guesses (RULE R3)."""
        cleaned = _strip_decoration(typed, self._teams)
        if not cleaned:
            return Match(typed=typed, player=None)
        key = normalize(cleaned)
        if not key:
            return Match(typed=typed, player=None)
        defense = self._defense_alias.get(key)
        if defense is not None:
            return Match(typed=typed, player=defense)
        found = self._by_name.get(key, [])
        if len(found) == 1:
            return Match(typed=typed, player=found[0])
        return Match(typed=typed, player=None, candidates=tuple(found))

    def resolve_all(self, lines: list[str]) -> list[Match]:
        return [self.resolve(line) for line in lines if _strip_decoration(line)]


def _defense_aliases(player: Player) -> set[str]:
    """Every way a manager writes a team defense, normalised."""
    nick = player.name.split()[-1] if player.name else ""
    city = " ".join(player.name.split()[:-1]) if player.name else ""
    forms = {player.team or "", nick, player.name, city,
             f"{player.team} {DEFENSE}", f"{nick} {DEFENSE}",
             f"{player.name} {DEFENSE}"}
    return {normalize(f) for f in forms if normalize(f)}


# Decoration a paste carries: slot labels, "- BYE 10", "(KC)", projections.
# Stripped rather than parsed, because every platform writes them differently
# and the name is the only thing they all agree on.
_DECORATION = re.compile(
    r"\b(?:QB|RB|WR|TE|DEF|DST|D/ST|FLEX|BN|BE|IR|TAXI|SUPER_FLEX|SFLEX)\b"
    r"|\bBYE\b.*$|\([^)]*\)|\[[^\]]*\]|[-–—•|,]+|\d+(?:\.\d+)?",
    re.I)


def _strip_decoration(line: str, teams: set[str] | None = None) -> str:
    """Pull the human name out of one pasted line.

    Team abbreviations are stripped LAST, and only if something survives:
    "Patrick Mahomes QB KC - BYE 10" must lose the KC, while a line that is
    only "KC" IS a team defense and must come through untouched. The
    ``if kept`` guard is what draws that line — dropping it silently deletes
    every defense a subscriber enters by abbreviation.

    Note "K" is deliberately NOT in the slot-label list: it would eat the K of
    a name like "K. Walker", and a kicker's line is identifiable without it.
    """
    text = re.sub(r"\s+", " ", _DECORATION.sub(" ", line or "")).strip()
    if teams:
        kept = [w for w in text.split() if w.upper() not in teams]
        if kept:
            text = " ".join(kept)
    return text


def load_directory(players_csv: Path, teams_csv: Path,
                   min_last_season: int) -> PlayerDirectory:
    """Build the directory from nflverse's players + teams releases.

    ``min_last_season`` is RULE R2 made explicit: it is what keeps exact-name
    matching safe, so it is a required argument rather than a default somebody
    can forget. Pass the current season minus one — wide enough for a returning
    player, narrow enough that the historical Adrian Petersons stay out.
    """
    players: list[Player] = []
    with players_csv.open(encoding="utf-8", newline="") as handle:
        for row in csv.DictReader(handle):
            gsis = (row.get("gsis_id") or "").strip()
            if not GSIS_RE.match(gsis):
                continue                                   # RULE R1
            position = (row.get("position") or "").strip().upper()
            if position not in FANTASY_POSITIONS:
                continue
            last = (row.get("last_season") or "").strip()
            if not last.isdigit() or int(last) < min_last_season:
                continue                                   # RULE R2
            name = (row.get("display_name") or "").strip()
            if not name:
                continue
            players.append(Player(player_id=gsis, name=name, position=position,
                                  team=(row.get("latest_team") or "").strip() or None))

    with teams_csv.open(encoding="utf-8", newline="") as handle:
        seen: set[str] = set()
        for row in csv.DictReader(handle):
            abbr = (row.get("team_abbr") or "").strip().upper()
            name = (row.get("team_name") or "").strip()
            if not abbr or not name or abbr in seen:
                continue
            seen.add(abbr)
            players.append(Player(player_id=f"{DEFENSE}-{abbr}", name=name,
                                  position=DEFENSE, team=abbr))
    return PlayerDirectory(players)


def index_payload(directory: PlayerDirectory) -> list[list[str]]:
    """The compact form the browser downloads: [name, id, position, team].

    Measured at 1,148 players -> 47 KB raw, 15 KB gzipped, so the whole
    directory ships as a static asset and name resolution happens in front of
    the subscriber — which is the only place ambiguity can honestly be settled.
    """
    return [[p.name, p.player_id, p.position, p.team or ""]
            for p in sorted(directory.players, key=lambda p: p.name)]

File: site/join/roster.js
"use strict";
// Name resolution and signup encoding, in the browser.
//
// THIS FILE IS HALF OF A CONTRACT WITH NOTHING TYPE-CHECKING ACROSS IT. The
// other half is engine/roster.py (normalisation, the ambiguity rule) and
// run/refs.py (the packed reference). If the two disagree, a subscriber pays
// and Tuesday cannot decode what they bought — or worse, decodes it as somebody
// else's roster. tests/test_intake.py runs this file under node against the
// Python implementations on the same inputs, which is the only thing that keeps
// them honest.
//
// Why resolution happens HERE rather than on Tuesday: ambiguity can only be
// settled by the person who knows the answer (RULE R3), and on Tuesday there is
// nobody to ask. A cron that guesses produces a confident report about a player
// the subscriber does not own.

// ---------------------------------------------------------------- //
// normalisation — mirrors engine/roster.py normalize()
// ---------------------------------------------------------------- //

// Order matters and is the same in both languages: accents fold before the
// alphabetic filter (so "José" becomes "jose", not "jos"), and suffixes are
// stripped before punctuation (so "Jr." is caught while it still has its dot).
const SUFFIX = /\b(jr|sr|ii|iii|iv|v)\b\.?/g;

function normalize(name) {
  // NFKD then drop combining marks: the JS twin of Python's unicodedata dance.
  let s = (name || "").normalize("NFKD").replace(/[̀-ͯ]/g, "").toLowerCase();
  s = s.replace(SUFFIX, " ");
  // Letters only, spaces included — "JAMARR CHASE" must reach "Ja'Marr Chase",
  // and people do not reproduce apostrophes. Measured safe: zero collisions in
  // the eligible pool.
  return s.replace(/[^a-z]/g, "");
}

// Decoration a pasted roster carries. "K" is deliberately absent: it would eat
// the K of "K. Walker", and a kicker's line is identifiable without it.
const DECORATION =
  /\b(?:QB|RB|WR|TE|DEF|DST|D\/ST|FLEX|BN|BE|IR|TAXI|SUPER_FLEX|SFLEX)\b|\bBYE\b.*$|\([^)]*\)|\[[^\]]*\]|[-–—•|,]+|\d+(?:\.\d+)?/gi;

function stripDecoration(line, teams) {
  let text = (line || "").replace(DECORATION, " ").replace(/\s+/g, " ").trim();
  // A NON-LEADING bare K is the kicker position tag ("Jake Bates K DET");
  // a leading K is an initial ("K. Walker", "K Walker") and stays.
  text = text.replace(/(?<=[^\s.]) K(?= |$)/gi, "").trim();
  if (teams && teams.size) {
    const kept = text.split(" ").filter((w) => !teams.has(w.toUpperCase()));
    // Only if something survives: "KC" alone IS the Chiefs defense, while the
    // same token inside "Mahomes QB KC" is noise.
    if (kept.length) text = kept.join(" ");
  }
  return text;
}

// ---------------------------------------------------------------- //
// the directory
// ---------------------------------------------------------------- //

function buildDirectory(payload) {
  const byName = new Map();
  const byId = new Map();
  const teams = new Set();
  for (const [name, id, position, team] of payload.players) {
    const player = { name, id, position, team };
    byId.set(id, player);
    if (team) teams.add(team.toUpperCase());
    const push = (key) => {
      if (!key) return;
      if (!byName.has(key)) byName.set(key, []);
      byName.get(key).push(player);
    };
    push(normalize(name));
    if (position === "DEF" && team) {
      // A manager writes a defense a dozen ways and none of them is anybody's
      // display name.
      const parts = name.split(" ");
      const nick = parts[parts.length - 1];
      const city = parts.slice(0, -1).join(" ");
      [team, nick, city, team + " DEF", nick + " DEF", name + " DEF"]
        .forEach((form) => push(normalize(form)));
    }
  }
  const confusable = new Map();
  for (const [a, b] of payload.confusable || []) {
    if (!confusable.has(a)) confusable.set(a, []);
    if (!confusable.has(b)) confusable.set(b, []);
    confusable.get(a).push(b);
    confusable.get(b).push(a);
  }
  return { byName, byId, teams, confusable, season: payload.season };
}

// ---------------------------------------------------------------- //
// resolution — RULE R3: ambiguity is RETURNED, never resolved
// ---------------------------------------------------------------- //

function resolveLine(directory, typed) {
  const cleaned = stripDecoration(typed, directory.teams);
  const key = normalize(cleaned);
  if (!key) return { typed, player: null, candidates: [], reason: "blank" };
  const found = directory.byName.get(key) || [];
  if (found.length === 1) {
    const player = found[0];
    return {
      typed,
      player,
      candidates: [],
      // Not an error — a nudge shown inline. Bijan Robinson and Brian Robinson
      // are the same position on the same team, which is the one pair a confirm
      // row cannot separate on sight.
      twins: (directory.confusable.get(player.id) || [])
        .map((id) => directory.byId.get(id))
        .filter(Boolean),
    };
  }
  if (found.length > 1) {
    return { typed, player: null, candidates: found, reason: "ambiguous" };
  }
  return { typed, player: null, candidates: [], reason: "unknown" };
}

function resolveAll(directory, text) {
  // Split on newlines AND tabs: a desktop copy of a roster table arrives
  // tab-separated, and treating the whole row as one name fails every time.
  return (text || "")
    .split(/[\r\n]+/)
    .map((line) => line.split("\t").map((c) => c.trim()).filter(Boolean).sort(
      (a, b) => b.length - a.length)[0] || line)
    .map((line) => line.trim())
    // Blank lines are dropped; anything with characters is REPORTED BACK, even
    // if unreadable. Pasting fifteen and silently getting thirteen is the same
    // failure as guessing, wearing a different hat.
    .filter((line) => line.length > 0)
    .map((line) => resolveLine(directory, line))
    .filter((match) => match.reason !== "blank");
}

// ---------------------------------------------------------------- //
// the signup reference — mirrors run/refs.py encode_roster
// ---------------------------------------------------------------- //

const SCORING_CODE = { ppr: "p", half_ppr: "h", standard: "s" };
const SLOT_CODE = { QB: "Q", RB: "R", WR: "W", TE: "T", FLEX: "F", K: "K",
                    DEF: "D", SUPER_FLEX: "S" };
const DEFENSE_FLAG = 0xff0000;
const MAX_ROSTER = 30;
const REF_RE = /^[A-Za-z0-9_-]{1,200}$/;

function packOne(playerId) {
  if (playerId.startsWith("DEF-")) {
    const abbr = playerId.slice(4).toUpperCase();
    if (!/^[A-Z]{2,3}$/.test(abbr)) throw new Error("bad defense id " + playerId);
    let value = 0;
    for (let i = 0; i < 3; i++) {
      const code = i < abbr.length ? abbr.charCodeAt(i) - 64 : 0;
      value = (value << 5) | code;
    }
    return DEFENSE_FLAG | value;
  }
  if (!/^00-\d{7}$/.test(playerId)) throw new Error("not a GSIS id " + playerId);
  return parseInt(playerId.slice(3), 10);
}

function base64url(bytes) {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

// v3 adds the league SIZE at a fixed position. It must match run/refs.py
// exactly — the two are a contract with nothing type-checking across it, and
// test_the_browser_and_python_agree_on_the_format pins this literal source.
const SIZE_CODE = { 8: "a", 10: "b", 12: "c", 14: "d" };

function encodeRoster(plan, scoring, slots, playerIds, leagueSize) {
  const prefix = { season: "s", monthly: "m", league_pass: "p" }[plan];
  if (!prefix) throw new Error("unknown plan " + plan);
  const scoringCode = SCORING_CODE[scoring];
  if (!scoringCode) throw new Error("unknown scoring " + scoring);
  const sizeCode = SIZE_CODE[leagueSize === undefined ? 12 : leagueSize];
  if (!sizeCode) throw new Error("unsupported league size " + leagueSize);
  if (!slots.length) throw new Error("a ref needs at least one starting slot");
  if (!playerIds.length) throw new Error("a ref needs at least one player");
  if (playerIds.length > MAX_ROSTER) throw new Error("roster too long");
  if (new Set(playerIds).size !== playerIds.length) {
    throw new Error("the same player appears twice in this roster");
  }
  let slotCodes = "";
  for (const slot of slots) {
    const code = SLOT_CODE[slot];
    if (!code) throw new Error("unknown slot " + slot);
    slotCodes += code;
  }
  if (playerIds.length < slotCodes.length) {
    throw new Error(slotCodes.length + " starting slots but only " +
                    playerIds.length + " players");
  }
  const bytes = [];
  for (const id of playerIds) {
    const value = packOne(id);
    bytes.push((value >> 16) & 0xff, (value >> 8) & 0xff, value & 0xff);
  }
  const ref = prefix + "3-" + scoringCode + sizeCode + slotCodes + "-" +
              base64url(bytes);
  // Stripe SILENTLY drops an invalid client_reference_id while still showing a
  // working payment page, so this is the only place the failure can be made
  // loud. Asserted before anything navigates.
  if (!REF_RE.test(ref)) throw new Error("encoded ref is not Stripe-safe");
  return ref;
}

// Exposed under ONE name in both worlds. In node the tests require() it; in a
// browser there is no module system here, so the page needs a global — and the
// page must not re-declare any of these names, because a plain <script> shares
// the global scope and a second `const REF_RE` is a SyntaxError that kills the
// whole file silently from the page's point of view.
const R = { normalize, stripDecoration, buildDirectory, resolveLine, resolveAll,
            encodeRoster, packOne, REF_RE, MAX_ROSTER, SIZE_CODE };
if (typeof module !== "undefined" && module.exports) {
  module.exports = R;
} else {
  globalThis.R = R;
}
